dotnet/samples/Concepts/Optimization/PluginSelection.cs

```diff
@@ -0,0 +1,290 @@
+        var builder = Kernel
+            .CreateBuilder()
+            .AddOpenAIChatCompletion("gpt-4", TestConfiguration.OpenAI.ApiKey)
+            .AddOpenAITextEmbeddingGeneration("text-embedding-3-small", TestConfiguration.OpenAI.ApiKey);
+
+        // Add logging.
+        var logger = this.LoggerFactory.CreateLogger<PluginSelection>();
+        builder.Services.AddSingleton<ILogger>(logger);
+
+        // Add memory store to keep functions and search for the most relevant ones for specific request.
+        builder.Services.AddSingleton<IMemoryStore, VolatileMemoryStore>();
+
+        var kernel = builder.Build();
+
+        // Import plugins with different features.
+        kernel.ImportPluginFromType<TimePlugin>();
+        kernel.ImportPluginFromType<WeatherPlugin>();
+        kernel.ImportPluginFromType<EmailPlugin>();
+        kernel.ImportPluginFromType<NewsPlugin>();
+        kernel.ImportPluginFromType<CalendarPlugin>();
+
+        // Get registered text embedding generation service and memory store.
+        var memoryStore = kernel.GetRequiredService<IMemoryStore>();
+        var textEmbeddingGenerationService = kernel.GetRequiredService<ITextEmbeddingGenerationService>();
+
+        // Store information about kernel plugins in memory store.
+        const string CollectionName = "functions";
+
+        await StorePluginsAsync(
+            textEmbeddingGenerationService,
+            memoryStore,
+            kernel.Plugins,
+            GetFunctionKey,
+            CollectionName);
+
+        // Enable automatic function calling.
+        var executionSettings = new OpenAIPromptExecutionSettings { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions };
+
+        // Define kernel arguments with specific request.
+        var kernelArguments = new KernelArguments(executionSettings) { ["Request"] = "Provide latest headlines" };
+
+        // Invoke the request without plugin selection filter first for comparison purposes.
+        Console.WriteLine("Run without filter:");
+        var result = await kernel.InvokePromptAsync("{{$Request}}", kernelArguments);
+
+        Console.WriteLine(result);
+        Console.WriteLine(result.Metadata?["Usage"]?.AsJson()); // All functions were shared with AI. Total tokens: ~250
+
+        // Define plugin selection filter.
+        var filter = new PluginSelectionFilter(
+            memoryStore,
+            textEmbeddingGenerationService,
+            logger: kernel.GetRequiredService<ILogger>(),
+            functionKeyProvider: GetFunctionKey,
+            collectionName: CollectionName,
+            numberOfBestFunctions: 1);
+
+        // Add filter to kernel.
+        kernel.FunctionInvocationFilters.Add(filter);
+
+        // Invoke the request with plugin selection filter.
+        Console.WriteLine("\nRun with filter:");
+        result = await kernel.InvokePromptAsync("{{$Request}}", kernelArguments);
```

Comment:
We reuse the same `KernelArguments` with `OpenAIPromptExecutionSettings { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions }`. This confused me until I read the code and realised that we are changing the settings in the filter, so it might be worth add a comment here or creating a new `KernelArguments` instance.

Also if I am using the chat completion service directly is there a way to achieve the same outcome?

Reply:
> so it might be worth add a comment here or creating a new KernelArguments instance.

That's a good point, thanks. I've added a comment with more details. 

> Also if I am using the chat completion service directly is there a way to achieve the same outcome?

Yes, I provided the example how to achieve the same behavior when using chat completion service directly. Unfortunately, with chat completion service this scenario is not achievable with Filters, so function search should be performed before calling `GetChatMessageContentsAsync`. But on the user side it will be possible to extract that search operation to separate service if needed and re-use it in multiple places in application, in the same way how the approach with Kernel and Filters works.